File: infra/pipeline_stack.py
"""CDK stack for the OneNote PDF Extractor automated pipeline."""

from constructs import Construct
import aws_cdk as cdk
from aws_cdk import (
    Stack,
    Duration,
    RemovalPolicy,
    aws_dynamodb as dynamodb,
    aws_events as events,
    aws_events_targets as targets,
    aws_iam as iam,
    aws_lambda as lambda_,
    aws_cloudwatch as cloudwatch,
    aws_s3 as s3,
    aws_ssm as ssm,
)


ENV_CONFIG = {
    "dev": {
        "schedule": "rate(1 day)",
        "lambda_memory": 1024,
        "lambda_timeout_minutes": 10,
        "removal_policy": RemovalPolicy.DESTROY,
        "cw_namespace": "OneNotePipeline/Dev",
    },
    "prod": {
        "schedule": "rate(6 hours)",
        "lambda_memory": 2048,
        "lambda_timeout_minutes": 15,
        "removal_policy": RemovalPolicy.RETAIN,
        "cw_namespace": "OneNotePipeline/Prod",
    },
}


class PipelineStack(Stack):
    """Stack defining all resources for the OneNote PDF Extractor pipeline."""

    def __init__(
        self, scope: Construct, construct_id: str, *, env_name: str = "dev", **kwargs
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        config = ENV_CONFIG.get(env_name, ENV_CONFIG["dev"])
        ssm_prefix = f"/onenote-pipeline/{env_name}"

        # --- S3 bucket for PDF storage ---
        pdf_bucket = s3.Bucket(
            self,
            "PdfStore",
            bucket_name=f"onenote-pdf-store-{env_name}-{self.account}",
            versioned=True,
            removal_policy=config["removal_policy"],
            auto_delete_objects=config["removal_policy"] == RemovalPolicy.DESTROY,
            encryption=s3.BucketEncryption.S3_MANAGED,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
        )

        # --- DynamoDB table for deduplication ---
        export_tracker = dynamodb.Table(
            self,
            "ExportTracker",
            table_name=f"ExportTracker-{env_name}",
            partition_key=dynamodb.Attribute(
                name="pageId", type=dynamodb.AttributeType.STRING
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            removal_policy=config["removal_policy"],
        )

        # --- SSM parameters (placeholders — actual values set manually) ---
        ssm_params = {}
        secure_params = [
            "ms-client-id",
            "ms-client-secret",
            "ms-tenant-id",
            "google-service-account-json",
            "notebooklm-project-id",
        ]
        for param_name in secure_params:
            ssm_params[param_name] = ssm.StringParameter(
                self,
                f"Param-{param_name}",
                parameter_name=f"{ssm_prefix}/{param_name}",
                string_value="PLACEHOLDER",
                description=f"Placeholder for {param_name} — replace with actual value via console or CLI",
                tier=ssm.ParameterTier.STANDARD,
            )

        ssm_params["section-id"] = ssm.StringParameter(
            self,
            "Param-section-id",
            parameter_name=f"{ssm_prefix}/section-id",
            string_value="PLACEHOLDER",
            description="OneNote section ID to export — replace with actual value",
            tier=ssm.ParameterTier.STANDARD,
        )

        # --- Lambda function (Java 21 managed runtime, fat JAR) ---
        pipeline_function = lambda_.Function(
            self,
            "PipelineFunction",
            function_name=f"onenote-pipeline-{env_name}",
            runtime=lambda_.Runtime.JAVA_21,
            handler="com.extractor.pipeline.PipelineHandler::handleRequest",
            code=lambda_.Code.from_asset("../target/onenote-pdf-extractor-1.0-SNAPSHOT.jar"),
            memory_size=config["lambda_memory"],
            timeout=Duration.minutes(config["lambda_timeout_minutes"]),
            reserved_concurrent_executions=1,
            environment={
                "SSM_PREFIX": ssm_prefix,
                "S3_BUCKET": pdf_bucket.bucket_name,
                "DYNAMO_TABLE": export_tracker.table_name,
                "CW_NAMESPACE": config["cw_namespace"],
                "CONCURRENCY_LIMIT": "5",
            },
            description="OneNote PDF Extractor automated pipeline",
        )

        # --- IAM: least-privilege permissions ---
        # DynamoDB read/write
        export_tracker.grant_read_write_data(pipeline_function)

        # S3 put (write only — no read/delete needed)
        pdf_bucket.grant_put(pipeline_function)

        # SSM get-parameter for all pipeline parameters
        pipeline_function.add_to_role_policy(
            iam.PolicyStatement(
                effect=iam.Effect.ALLOW,
                actions=["ssm:GetParameter"],
                resources=[
                    f"arn:aws:ssm:{self.region}:{self.account}:parameter{ssm_prefix}/*"
                ],
            )
        )

        # CloudWatch put-metric-data
        pipeline_function.add_to_role_policy(
            iam.PolicyStatement(
                effect=iam.Effect.ALLOW,
                actions=["cloudwatch:PutMetricData"],
                resources=["*"],
                conditions={
                    "StringEquals": {
                        "cloudwatch:namespace": config["cw_namespace"]
                    }
                },
            )
        )

        # --- EventBridge scheduled rule ---
        schedule_expression = events.Schedule.expression(config["schedule"])

        rule = events.Rule(
            self,
            "PipelineSchedule",
            rule_name=f"onenote-pipeline-schedule-{env_name}",
            schedule=schedule_expression,
            description=f"Triggers OneNote pipeline ({env_name})",
        )
        rule.add_target(targets.LambdaFunction(pipeline_function))

        # --- CloudWatch alarm on Lambda errors ---
        error_metric = pipeline_function.metric_errors(
            period=Duration.minutes(5),
            statistic="Sum",
        )

        alarm = cloudwatch.Alarm(
            self,
            "PipelineErrorAlarm",
            alarm_name=f"onenote-pipeline-errors-{env_name}",
            metric=error_metric,
            threshold=1,
            evaluation_periods=1,
            comparison_operator=cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
            alarm_description=f"Alarm when OneNote pipeline Lambda errors >= 1 ({env_name})",
            treat_missing_data=cloudwatch.TreatMissingData.NOT_BREACHING,
        )

        # --- Outputs ---
        cdk.CfnOutput(self, "PdfBucketName", value=pdf_bucket.bucket_name)
        cdk.CfnOutput(self, "ExportTrackerTableName", value=export_tracker.table_name)
        cdk.CfnOutput(self, "LambdaFunctionName", value=pipeline_function.function_name)
        cdk.CfnOutput(self, "ScheduleRuleName", value=rule.rule_name)
        cdk.CfnOutput(self, "ErrorAlarmName", value=alarm.alarm_name)

File: infra/lib/pipeline-stack.ts
import { Construct } from "constructs";
import * as cdk from "aws-cdk-lib";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as events from "aws-cdk-lib/aws-events";
import * as targets from "aws-cdk-lib/aws-events-targets";
import * as iam from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as ssm from "aws-cdk-lib/aws-ssm";

interface EnvConfig {
  schedule: string;
  lambdaMemory: number;
  lambdaTimeoutMinutes: number;
  removalPolicy: cdk.RemovalPolicy;
  cwNamespace: string;
}

const ENV_CONFIG: Record<string, EnvConfig> = {
  dev: {
    schedule: "rate(1 day)",
    lambdaMemory: 1024,
    lambdaTimeoutMinutes: 10,
    removalPolicy: cdk.RemovalPolicy.DESTROY,
    cwNamespace: "OneNotePipeline/Dev",
  },
  prod: {
    schedule: "rate(6 hours)",
    lambdaMemory: 2048,
    lambdaTimeoutMinutes: 15,
    removalPolicy: cdk.RemovalPolicy.RETAIN,
    cwNamespace: "OneNotePipeline/Prod",
  },
};

interface PipelineStackProps extends cdk.StackProps {
  envName: string;
}

export class PipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: PipelineStackProps) {
    super(scope, id, props);

    const config = ENV_CONFIG[props.envName] ?? ENV_CONFIG["dev"];
    const ssmPrefix = `/onenote-pipeline/${props.envName}`;

    // --- S3 bucket for PDF storage ---
    const pdfBucket = new s3.Bucket(this, "PdfStore", {
      bucketName: `onenote-pdf-store-${props.envName}-${this.account}`,
      versioned: true,
      removalPolicy: config.removalPolicy,
      autoDeleteObjects: config.removalPolicy === cdk.RemovalPolicy.DESTROY,
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    });

    // --- DynamoDB table for deduplication ---
    const exportTracker = new dynamodb.Table(this, "ExportTracker", {
      tableName: `ExportTracker-${props.envName}`,
      partitionKey: { name: "pageId", type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      removalPolicy: config.removalPolicy,
    });

    // --- SSM parameters (placeholders — actual values set manually) ---
    const secureParams = [
      "ms-client-id",
      "ms-client-secret",
      "ms-tenant-id",
      "google-service-account-json",
      "notebooklm-project-id",
    ];
    for (const paramName of secureParams) {
      new ssm.StringParameter(this, `Param-${paramName}`, {
        parameterName: `${ssmPrefix}/${paramName}`,
        stringValue: "PLACEHOLDER",
        description: `Placeholder for ${paramName} — replace with actual value via console or CLI`,
        tier: ssm.ParameterTier.STANDARD,
      });
    }

    new ssm.StringParameter(this, "Param-section-id", {
      parameterName: `${ssmPrefix}/section-id`,
      stringValue: "PLACEHOLDER",
      description: "OneNote section ID to export — replace with actual value",
      tier: ssm.ParameterTier.STANDARD,
    });

    // --- Lambda function (Java 21 managed runtime, fat JAR) ---
    const pipelineFunction = new lambda.Function(this, "PipelineFunction", {
      functionName: `onenote-pipeline-${props.envName}`,
      runtime: lambda.Runtime.JAVA_21,
      handler: "com.extractor.pipeline.PipelineHandler::handleRequest",
      code: lambda.Code.fromAsset("../target/onenote-pdf-extractor-1.0-SNAPSHOT.jar"),
      memorySize: config.lambdaMemory,
      timeout: cdk.Duration.minutes(config.lambdaTimeoutMinutes),
      reservedConcurrentExecutions: 1,
      environment: {
        SSM_PREFIX: ssmPrefix,
        S3_BUCKET: pdfBucket.bucketName,
        DYNAMO_TABLE: exportTracker.tableName,
        CW_NAMESPACE: config.cwNamespace,
        CONCURRENCY_LIMIT: "5",
      },
      description: "OneNote PDF Extractor automated pipeline",
    });

    // --- IAM: least-privilege permissions ---
    exportTracker.grantReadWriteData(pipelineFunction);
    pdfBucket.grantPut(pipelineFunction);

    pipelineFunction.addToRolePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["ssm:GetParameter"],
        resources: [
          `arn:aws:ssm:${this.region}:${this.account}:parameter${ssmPrefix}/*`,
        ],
      })
    );

    pipelineFunction.addToRolePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["cloudwatch:PutMetricData"],
        resources: ["*"],
        conditions: {
          StringEquals: { "cloudwatch:namespace": config.cwNamespace },
        },
      })
    );

    // --- EventBridge scheduled rule ---
    const rule = new events.Rule(this, "PipelineSchedule", {
      ruleName: `onenote-pipeline-schedule-${props.envName}`,
      schedule: events.Schedule.expression(config.schedule),
      description: `Triggers OneNote pipeline (${props.envName})`,
    });
    rule.addTarget(new targets.LambdaFunction(pipelineFunction));

    // --- CloudWatch alarm on Lambda errors ---
    const errorMetric = pipelineFunction.metricErrors({
      period: cdk.Duration.minutes(5),
      statistic: "Sum",
    });

    new cloudwatch.Alarm(this, "PipelineErrorAlarm", {
      alarmName: `onenote-pipeline-errors-${props.envName}`,
      metric: errorMetric,
      threshold: 1,
      evaluationPeriods: 1,
      comparisonOperator:
        cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      alarmDescription: `Alarm when OneNote pipeline Lambda errors >= 1 (${props.envName})`,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    });

    // --- Outputs ---
    new cdk.CfnOutput(this, "PdfBucketName", { value: pdfBucket.bucketName });
    new cdk.CfnOutput(this, "ExportTrackerTableName", { value: exportTracker.tableName });
    new cdk.CfnOutput(this, "LambdaFunctionName", { value: pipelineFunction.functionName });
    new cdk.CfnOutput(this, "ScheduleRuleName", { value: rule.ruleName });
    new cdk.CfnOutput(this, "ErrorAlarmName", { value: `onenote-pipeline-errors-${props.envName}` });
  }
}
